### PR TITLE
Add shared TestLoggerListener and remove redundant

### PR DESCRIPTION
**Pull request by hellosnow**

### CODE REVIEW

File: test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs [hunk 1]
```diff
@@ -15,22 +15,29 @@ public class TestLoggerListener : ILoggerListener
...
         public void WriteLine(ILogItem item)
         {
-            if (Phase == null || item.Phase == Phase)
+            if (item == null || item.LogLevel < LogLevelThreshold)
+            {
+                return;
+            }
+            if (Phase == null ||
+                item.Phase == Phase ||
+                !EnablePhaseEndWith && item.Phase.StartsWith(Phase) ||
```
Comment (vicancy) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: >EnablePhaseEndWith [](start = 17, length = 18)

EnableFuzzyMatch and use contains?

Comment (vicancy) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: Or a Match delegate?

---
In reply to: [119277973](https://github.com/dotnet/docfx/pull/1712#discussion_r119277973) [](ancestors = 119277973)

File: test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs [hunk 2]
```diff
@@ -15,22 +15,29 @@ public class TestLoggerListener : ILoggerListener
...
-        }
+        public bool EnablePhaseEndWith { get; set; }
 
-        public void Dispose()
+        public TestLoggerListener()
         {
         }
 
-        public void Flush()
+        public TestLoggerListener(string phase, LogLevel logLevelThreshold = LogLevel.Warning, bool enablePhaseEndWith = false)
```
Comment (vicancy) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: >phase [](start = 41, length = 5)

phase = null?

File: test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs [hunk 3]
```diff
@@ -15,22 +15,29 @@ public class TestLoggerListener : ILoggerListener
...
 
         public void WriteLine(ILogItem item)
         {
-            if (Phase == null || item.Phase == Phase)
+            if (item == null || item.LogLevel < LogLevelThreshold)
+            {
+                return;
+            }
+            if (Phase == null ||
+                item.Phase == Phase ||
```
Comment (vicancy) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: >== [](start = 27, length = 2)

ignore case allowed?

File: test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs [hunk 4]
```diff
@@ -15,22 +16,28 @@ public class TestLoggerListener : ILoggerListener
...
+            Func<ILogItem, bool> phaseMatcher = null)
         {
-        }
+            Phase = phase;
+            LogLevelThreshold = logLevelThreshold;
 
-        public void Flush()
-        {
+            // Set default phase matcher to start with current phase
+            PhaseMatcher = phaseMatcher ?? (iLogItem => iLogItem?.Phase != null && iLogItem.Phase.StartsWith(Phase));
```
Comment (vicancy) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: >Phase [](start = 109, length = 5)

could be null?

**Comment by superyyrrzz** [comment]
:shipit:

File: test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs (hunk 4 above)
Comment (hellosnow) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: refer to line 38, if phase is null, the item will be added to items, need not check again here?

---
In reply to: [119304710](https://github.com/dotnet/docfx/pull/1712#discussion_r119304710) [](ancestors = 119304710)

File: test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs [hunk 5]
```diff
@@ -15,22 +16,28 @@ public class TestLoggerListener : ILoggerListener
...
         public LogLevel LogLevelThreshold { get; set; }
 
-        public TestLoggerListener(string phase = null)
-        {
-            Phase = phase;
-        }
+        public Func<ILogItem, bool> PhaseMatcher { get; set; }
 
-        public void Dispose()
+        public TestLoggerListener(string phase = null,
```
Comment (vwxyzh) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: use create method?

File: test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    using System;
+    using System.Collections.Generic;
+
+    using Microsoft.DocAsCode.Common;
+
+    public class TestLoggerListener : ILoggerListener
+    {
+        public List<ILogItem> Items { get; } = new List<ILogItem>();
+
+        private Func<ILogItem, bool> _matcher;
```
Comment (vwxyzh) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: >  [](start = 15, length = 1)

readonly

File: test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        public static TestLoggerListener CreateLoggerListenerWithPhaseStartMatcher(string phase, LogLevel logLevel = LogLevel.Warning)
+        {
+            return new TestLoggerListener(iLogItem =>
+            {
+                if (iLogItem.LogLevel < logLevel)
+                {
+                    return false;
+                }
+                if (phase == null ||
+                   (iLogItem?.Phase != null && iLogItem.Phase.StartsWith(phase, StringComparison.OrdinalIgnoreCase)))
```
Comment (vwxyzh) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: >OrdinalIgnoreCase [](start = 97, length = 17)

exactly match?

File: test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+    using Microsoft.DocAsCode.Common;
+
+    public class TestLoggerListener : ILoggerListener
+    {
+        public List<ILogItem> Items { get; } = new List<ILogItem>();
+
+        private Func<ILogItem, bool> _matcher;
+
+        private TestLoggerListener(Func<ILogItem, bool> matcher)
```
Comment (vwxyzh) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: >private [](start = 8, length = 7)

public

Comment (vwxyzh) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: >matcher [](start = 56, length = 7)

filter?

**Comment by vwxyzh** [comment]
:shipit:

File: test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs (hunk 7 above)
Comment (hellosnow) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: discuss offline, the comparision should be case sensitive.

---
In reply to: [119785752](https://github.com/dotnet/docfx/pull/1712#discussion_r119785752) [](ancestors = 119785752)

File: test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+                }
+                return false;
+            });
+        }
+
+        #endregion
+
+        #region Helpers 
+
+        public ILogItem TakeAndRemove()
```
Comment (vicancy) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: >TakeAndRemove [](start = 24, length = 13)

when is it used? is thread safe important?

Comment (hellosnow) [inline test/Microsoft.DocAsCode.Tests.Common/TestLoggerListener.cs]: it's used for specific case, will remove and not common for the listener.

---
In reply to: [119790293](https://github.com/dotnet/docfx/pull/1712#discussion_r119790293) [](ancestors = 119790293)

**Comment by vicancy** [comment]
:shipit:

**Pull request merged**